### PR TITLE
Create new grpc storage configuration to align with OTEL

### PR DESCRIPTION
**Pull request by akagami-harsh**
## Which problem is this PR solving?
- part of #5229 

## Description of the changes
- added more grpc storage client configuration to align with otel
 ```
	configgrpc.ClientConfig        `mapstructure:",squash"`
	exporterhelper.TimeoutSettings `mapstructure:",squash"`
 ```
These are not all the configs, but i'll add more based on feedback on this initial approach.

## How was this change tested?
- not tested yet 

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -151,7 +151,7 @@ require (
 	github.com/mitchellh/reflectwalk v1.0.2 // indirect
 	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
 	github.com/modern-go/reflect2 v1.0.2 // indirect
-	github.com/mostynb/go-grpc-compression v1.2.2 // indirect
+	github.com/mostynb/go-grpc-compression v1.2.2
```
Comment (yurishkuro) [inline go.mod:154]: this seems unrelated to this PR

File: plugin/storage/grpc/config/config.go [hunk 2]
```diff
@@ -90,7 +100,28 @@ func (c *Configuration) buildRemote(logger *zap.Logger, tracerProvider trace.Tra
 	opts := []grpc.DialOption{
 		grpc.WithStatsHandler(otelgrpc.NewClientHandler(otelgrpc.WithTracerProvider(tracerProvider))),
 		grpc.WithBlock(),
+		grpc.WithDefaultCallOptions(grpc.WaitForReady(c.WaitForReady)),
+	}
+
+	if c.Compression.IsCompressed() {
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:106]: Doesn't OTEL provide a utility function in their config packages that helps with translating gRPC server config into settings? We should avoid reinventing the wheel and just reuse what they have.

Comment (akagami-harsh) [reply plugin/storage/grpc/config/config.go]: To reuse that, we need to fully utilize all the features it has. We can do this for every feature except `auth` (authentication) , since it needs to be supported by both the client and server. and also the `auth` is tightly coupled with auth extension https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/configauth

Comment (akagami-harsh) [reply plugin/storage/grpc/config/config.go]: > Do we need to do something special for auth?

yes, i think from the client side we can reuse the otel's authentication extensions but, from the server side we need to find a find a way to support authentication.

File: plugin/storage/grpc/config/config.go [hunk 3]
```diff
@@ -47,10 +50,11 @@ type Configuration struct {
...
 
-	pluginHealthCheck     *time.Ticker
-	pluginHealthCheckDone chan bool
-	pluginRPCClient       plugin.ClientProtocol
-	remoteConn            *grpc.ClientConn
+	pluginHealthCheck       *time.Ticker
+	pluginHealthCheckDone   chan bool
+	pluginRPCClient         plugin.ClientProtocol
+	remoteConn              *grpc.ClientConn
+	configgrpc.ClientConfig `mapstructure:",squash"`
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:57]: Let's create a new config struct like ConfigV2 and only keep the fields that are really needed there. For instance, we do not plan to support sidecar plugin model, so the first three fields at the top are not needed.

Also, I assume otel ClientConfig would already include things like TLS and timeout and address, so fields 4-6 also should not be needed in the new config.

Basically, I would suggest this refactoring:
  * add ConfigV2 with only what we need
  * refactor remote storage initialization code to use OTEL helpers for interpreting ClientConfig into a gRPC connection
  * write a translator from v1 Configuration into ConfigV2 (V2 should be a superset, except for the sidecar mode).

Comment (akagami-harsh) [reply plugin/storage/grpc/config/config.go]: now the sidecar plugin config fields are removed, do we still need to create a new configV2 and a translator?

File: cmd/remote-storage/app/server.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ import (
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
+	_ "google.golang.org/grpc/encoding"
```
Comment (yurishkuro) [inline cmd/remote-storage/app/server.go:25]: Is this the encoding we need, or the gogo one from pkg/ ?

File: plugin/storage/grpc/config/config.go [hunk 5]
```diff
@@ -41,6 +46,30 @@ type Configuration struct {
 	remoteConn *grpc.ClientConn
 }
 
+type ConfigV2 struct {
+	Configuration `mapstructure:",squash"`
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:50]: The idea of cfg v2 is to have a clean break, not to keep using v1 config (ok to use it internally when starting the implementation, but not as external config struct)

File: plugin/storage/grpc/factory.go [hunk 6]
```diff
@@ -62,7 +62,7 @@ func NewFactory() *Factory {
 
 // NewFactoryWithConfig is used from jaeger(v2).
 func NewFactoryWithConfig(
-	cfg config.Configuration,
+	cfg config.ConfigV2,
```
Comment (yurishkuro) [inline plugin/storage/grpc/factory.go:65]: +1

File: cmd/remote-storage/app/server.go (hunk 4 above)
Comment (akagami-harsh) [reply cmd/remote-storage/app/server.go]: earlier i added it because i was getting a warning when used the `Compression` config from `configgrpc.ClientConfig` . but now i am not getting any such warning. so removed it for now

File: plugin/storage/grpc/config/config.go [hunk 7]
```diff
@@ -37,10 +42,34 @@ type Configuration struct {
 	RemoteTLS            tlscfg.Options
 	RemoteConnectTimeout time.Duration `yaml:"connection-timeout" mapstructure:"connection-timeout"`
 	TenancyOpts          tenancy.Options
+}
+
+type ConfigV2 struct {
+	Configuration
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:48]: You are still redefining v2 in terms of v1 - you need to break v1 dependency completely (super easy in case of this specific storage since v1 config is quire small). Otherwise you're just duplicating the settings, e.g. tenancy and TLS are also defined in v1 Configuration

File: plugin/storage/grpc/config/config.go [hunk 8]
```diff
@@ -37,10 +42,34 @@ type Configuration struct {
...
+	TenancyOpts                    tenancy.Options
+	configgrpc.ClientConfig        `mapstructure:",squash"`
+	exporterhelper.TimeoutSettings `mapstructure:",squash"`
 
 	remoteConn *grpc.ClientConn
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:53]: can we avoid having runtime objects in the configuration structs? It was a bad pattern in v1

File: plugin/storage/grpc/config/config.go [hunk 9]
```diff
@@ -55,17 +84,21 @@ type PluginBuilder interface {
 }
 
 // Build instantiates a PluginServices
-func (c *Configuration) Build(logger *zap.Logger, tracerProvider trace.TracerProvider) (*ClientPluginServices, error) {
-	return c.buildRemote(logger, tracerProvider, grpc.NewClient)
+func (c *ConfigV2) Build(logger *zap.Logger, tracerProvider trace.TracerProvider) (*ClientPluginServices, error) {
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:87]: I suggest 
```
func NewRemoteStorage(*ConfigV2)
```

leave Build where it is, with Configuration, but re-implement as
```
func (c *Configuration) Build(...) {
    v2cfg = c.AsV2()
    s := NewRemoteStorage(v2cfg)
}
```

Comment (akagami-harsh) [reply plugin/storage/grpc/config/config.go]: PTAL, is this what you intended?

File: plugin/storage/grpc/config/config.go (hunk 8 above)
Comment (akagami-harsh) [reply plugin/storage/grpc/config/config.go]: moved it in `factory` struct

File: plugin/storage/grpc/config/config.go [hunk 10]
```diff
@@ -37,8 +39,32 @@ type Configuration struct {
...
+	exporterhelper.TimeoutSettings `mapstructure:",squash"`
+}
+
+func (c *Configuration) translateToConfigV2() *ConfigV2 {
+	v2 := &ConfigV2{}
+	v2.Endpoint = c.RemoteServerAddr
+	v2.Timeout = c.RemoteConnectTimeout
 
-	remoteConn *grpc.ClientConn
+	v2.ClientConfig.TLSSetting.Insecure = !c.RemoteTLS.Enabled
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:55]: let's move all these assignments into a method on the `tlscfg` object
```
func (o *Options) ToOtelClientConfig()
```

File: plugin/storage/grpc/config/config.go [hunk 11]
```diff
@@ -50,60 +76,53 @@ type ClientPluginServices struct {
...
 // Build instantiates a PluginServices
-func (c *Configuration) Build(logger *zap.Logger, tracerProvider trace.TracerProvider) (*ClientPluginServices, error) {
-	return c.buildRemote(logger, tracerProvider, grpc.NewClient)
+func (c *Configuration) Build(logger *zap.Logger, tracerProvider trace.TracerProvider) (*ClientPluginServices, *grpc.ClientConn, error) {
+	v2Cfg := c.translateToConfigV2()
+	s, remoteConn, err := newRemoteStorage(v2Cfg, tracerProvider)
+	if err != nil {
+		return nil, nil, err
+	}
+	return s, remoteConn, nil
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:89]:
```suggestion
	return v2Cfg.Build(...)
```

File: plugin/storage/grpc/config/config.go [hunk 12]
```diff
@@ -50,60 +76,53 @@ type ClientPluginServices struct {
...
 
 	tenancyMgr := tenancy.NewManager(&c.TenancyOpts)
 	if tenancyMgr.Enabled {
 		opts = append(opts, grpc.WithUnaryInterceptor(tenancy.NewClientUnaryInterceptor(tenancyMgr)))
 		opts = append(opts, grpc.WithStreamInterceptor(tenancy.NewClientStreamInterceptor(tenancyMgr)))
 	}
-	var err error
-	c.remoteConn, err = newClient(c.RemoteServerAddr, opts...)
+
+	remoteConn, err := c.ToClientConn(context.Background(), componenttest.NewNopHost(), component.TelemetrySettings{}, opts...)
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:115]: > `componenttest.NewNopHost()`, `component.TelemetrySettings{}`

that doesn't make sense in production code, if these args are required we need to somehow pass them in as arguments.

File: plugin/storage/grpc/config/config.go [hunk 13]
```diff
@@ -50,60 +76,53 @@ type ClientPluginServices struct {
...
+}
 
-func (c *Configuration) buildRemote(logger *zap.Logger, tracerProvider trace.TracerProvider, newClient newClientFn) (*ClientPluginServices, error) {
+func newRemoteStorage(c *ConfigV2, tracerProvider trace.TracerProvider) (*ClientPluginServices, *grpc.ClientConn, error) {
 	opts := []grpc.DialOption{
 		grpc.WithStatsHandler(otelgrpc.NewClientHandler(otelgrpc.WithTracerProvider(tracerProvider))),
 		grpc.WithBlock(),
 	}
-	if c.RemoteTLS.Enabled {
-		tlsCfg, err := c.RemoteTLS.Config(logger)
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:70]: This used to create a background thread that was watching for certificate file changes. How is this done in OTEL? It won't be an equivalent change if we don't support cert reloads.

File: plugin/storage/grpc/factory.go [hunk 14]
```diff
@@ -92,14 +95,17 @@ func (f *Factory) InitFromViper(v *viper.Viper, logger *zap.Logger) {
 func (f *Factory) configureFromOptions(opts Options) {
 	f.options = opts
 	f.builder = &f.options.Configuration
+	f.builder = &f.options.ConfigV2
```
Comment (yurishkuro) [inline plugin/storage/grpc/factory.go:98]: why have `options.Configuration` if we won't use it?

File: plugin/storage/grpc/config/config.go [hunk 15]
```diff
@@ -50,60 +76,53 @@ type ClientPluginServices struct {
...
+func (c *ConfigV2) Build(logger *zap.Logger, tracerProvider trace.TracerProvider) (*ClientPluginServices, *grpc.ClientConn, error) {
+	s, remoteConn, err := newRemoteStorage(c, tracerProvider)
+	if err != nil {
+		return nil, nil, err
+	}
+	return s, remoteConn, nil
+}
 
-func (c *Configuration) buildRemote(logger *zap.Logger, tracerProvider trace.TracerProvider, newClient newClientFn) (*ClientPluginServices, error) {
+func newRemoteStorage(c *ConfigV2, tracerProvider trace.TracerProvider) (*ClientPluginServices, *grpc.ClientConn, error) {
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:100]: Do yo think it's actually important to the caller of this function to get back `*grpc.ClientConn`? I understand that you are returning it in order to ensure it's closed in the end, but from the caller perspective what it's trying to do is instantiate `*ClientPluginServices`. The connection is just an implementation detail of `*ClientPluginServices`, the caller does not need to know about it. Instead we can capture the connection inside `*ClientPluginServices` and have `*ClientPluginServices` implement a Close method.

File: plugin/storage/grpc/options.go [hunk 16]
```diff
@@ -37,6 +37,7 @@ const (
 // to bind them to command line flags
 type Options struct {
 	Configuration config.Configuration `mapstructure:",squash"`
+	ConfigV2      config.ConfigV2      `mapstructure:",squash"`
```
Comment (yurishkuro) [inline plugin/storage/grpc/options.go:40]: I recommend leaving Options only as a container for Configuration. It's still needed to support v1 CLI parsing, but in the factory when time comes to start creating services we can convert it to ConfigV2 once.

**Review by yurishkuro — COMMENTED** [review]
it looks like you're picking some go.mod changes because your branch is not up to date. Please rebase on main

**Review by yurishkuro — COMMENTED** [review]
I refactored & cleaned-up a bit more to make the code simpler.

File: pkg/config/tlscfg/options.go [hunk 17]
```diff
@@ -135,6 +136,23 @@ func (p Options) loadCertPool() (*x509.CertPool, error) {
 	return certPool, nil
 }
 
+func (o *Options) ToOtelClientConfig() configtls.ClientConfig {
```
Comment (yurishkuro) [inline pkg/config/tlscfg/options.go:139]: we should add a test for this to keep the coverage up

File: plugin/storage/grpc/config/config.go [hunk 18]
```diff
@@ -37,72 +39,80 @@ type Configuration struct {
 	RemoteTLS            tlscfg.Options
 	RemoteConnectTimeout time.Duration `yaml:"connection-timeout" mapstructure:"connection-timeout"`
 	TenancyOpts          tenancy.Options
+}
+
+type ConfigV2 struct {
+	TenancyOpts                    tenancy.Options
```
Comment (yurishkuro) [inline plugin/storage/grpc/config/config.go:45]:
```suggestion
	Tenancy                    tenancy.Options
```
(why would we want "Opts" suffix in the config files?)

**Review by yurishkuro — APPROVED** [review]
